### ISSUE TITLE
bug(core): viewRef._viewContainerRef can leak memory (ViewRef, ViewContainerRef)

### ISSUE THREAD

**Issue by Achilles1515** [angular/angular#38648]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/core

### Is this a regression?

This is a problem in Ivy and View Engine.

### Description

The `attachToViewContainerRef()` method of the `ViewRef` class is called whenever the `ViewRef` is attached to a `ViewContainerRef`, and this `ViewContainerRef` instance is assigned to `viewRef._viewContainerRef`.

https://github.com/angular/angular/blob/930eeaf177a4c277f437f42314605ff8dc56fc82/packages/core/src/render3/view_ref.ts#L284-L289

The **only** place where the `_viewContainerRef` property is cleared is when `viewRef.destroy()` is manually called.

https://github.com/angular/angular/blob/930eeaf177a4c277f437f42314605ff8dc56fc82/packages/core/src/render3/view_ref.ts#L75-L88

When `viewContainerRef.detach()` is used to detach a `ViewRef`, the `_viewContainerRef` property is still referencing the container from which it was detached. If this container should be destroyed (e.g. a `ViewContainerRef` of a component being destroyed), then the `_viewContainerRef` property will continue to hold a reference to the destroyed container, causing a memory leak (until the next time this `ViewRef` is attached to a different container).

----

Upon initial glance of the source code, I was not able to find a quick solution to this issue. I don't know if there is a way to access a `ViewRef` instance inside the node_manipulation `detachView()` function.

https://github.com/angular/angular/blob/930eeaf177a4c277f437f42314605ff8dc56fc82/packages/core/src/render3/node_manipulation.ts#L319

Essentially, there needs to be some type of `detachFromViewContainerRef()` method to complement the `attachToViewContainerRef()` method.

## 🔬 Minimal Reproduction

[StackBlitz example](https://stackblitz.com/edit/angular-ymlxv8?file=src%2Fapp%2Fapp.component.ts)
What's happening here:
1. Manually create an `EmbeddedViewRef`.
2. Manually create a dynamic component.
3. Attach the `EmbeddedViewRef` to the dynamic component's `ViewContainerRef`.
4. Detach the `EmbeddedViewRef`.
5. Destroy the dynamic component.

----

(Using Chrome DevTools)
- Load the StackBlitz example.
- Take heap snapshot before doing anything.
	- Search objects for `ViewContainerRef` and you'll find **two** instances.
- Click "create and attach" button.
- Take heap snapshot.
	- Search objects for `ViewContainerRef` and you'll find **three** instances.
- Click "detach and destroy" button.
- Take heap snapshot.
	- Search objects for `ViewContainerRef` and you'll find **three** instances.
	- MEMORY LEAK --> the third `ViewContainerRef` should be garbage collected, as there are no references to it being held in user code

Now uncomment line 35 in app.component.ts that manually clears the `_viewContainerRef` property on `ViewRef` detach:

```ts
// Manually clear the view ref's assigned previously assigned view container
// --> Allows the `viewContainerRef` to be garbage collected.
(this.viewRef as any)._viewContainerRef = null;
```

Do the same steps as above.
After clicking on "detach and destroy", there will only be **two** `ViewContainerRef` instances which is expected behavior.

## 🌍  Your Environment

**Angular Version:**
10.0.14

**Comment by dmk1111** [comment]
As far as I see from code, this is expected behavior:
https://github.com/angular/angular/blob/d1ea1f4c7f3358b730b0d94e65b00bc28cae279c/packages/core/src/linker/view_container_ref.ts#L139-L145

So you have to destroy viewRef calling `destroy` instead of `detach`

**Comment by Achilles1515** [comment]
> Correct me if I'm wrong, but as far as I see from code, this is expected behavior:
> So you have to destroy viewRef calling `destroy` instead of `detach`

I'm not really talking about destroying a `ViewRef`. Anytime a `ViewRef` is removed from a `ViewRefContainer` (other than using `viewRef.destory()`), `viewRef._viewContainerRef ` still points to that `ViewRefContainer` instance. Therefore that `ViewRefContainer` will never be garbage collected until the `ViewRef` instance is attached to another container or its `destroy` method is called.

**Comment by Achilles1515** [comment]
@pkozlowski-opensource can you confirm this issue?

**Comment by jelbourn** [comment]
Looked at this with @mhevery, sounds like a legitimate bug. Summary is:
* `ViewRef._viewContainerRef` references a `ViewContainerRef` 
* This property is never cleared when the `ViewRef` is _detached_ from that `ViewContainerRef`
* Unclear if this property is unset when the `ViewContainerRef` is destroyed
